In [57]:
import pandas as pd
import seaborn as sns
import numpy as np

In [58]:
titanic_df = pd.read_csv('./datasets/titanic.csv')
titanic_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [59]:
# find datatype of each column.

In [60]:
titanic_df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [61]:
type(titanic_df)

pandas.core.frame.DataFrame

In [62]:
# get initial insights from data

In [63]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In Above description,

    - Count shows Age column is less than all other column.
    - mean shows the average. average for passenger id is not significant (we can ignore passenger id mean here).
    - Average fare paid by passenger is 32.20
    - Average age of person is 29.
    - Minimum fare is 0, min age is 0.42, 

## Percentiles

25th percentil, 50th percentile and 75th Percentile?

In [64]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Analyzing Dataset 

- Cabin column has more nulls Since there are 204 null values.
- 

In [65]:
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [66]:
titanic_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### filter dataset whose datatype is object

In [67]:
# extract columns with object as datatype. 
filtered_cols = titanic_df.dtypes[titanic_df.dtypes=='object'].index
filtered_cols

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [68]:
titanic_df[filtered_cols].head()

Name     Sex  \
0                            Braund, Mr. Owen Harris    male   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                             Heikkinen, Miss. Laina  female   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                           Allen, Mr. William Henry    male   

             Ticket Cabin Embarked  
0         A/5 21171   NaN        S  
1          PC 17599   C85        C  
2  STON/O2. 3101282   NaN        S  
3            113803  C123        S  
4            373450   NaN        S

In [69]:
titanic_df[filtered_cols].describe()

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

In [70]:
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [71]:
# get name column - get first 15 records

In [72]:
titanic_df['Name'][0:15]

0                               Braund, Mr. Owen Harris
1     Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                Heikkinen, Miss. Laina
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                              Allen, Mr. William Henry
5                                      Moran, Mr. James
6                               McCarthy, Mr. Timothy J
7                        Palsson, Master. Gosta Leonard
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                   Nasser, Mrs. Nicholas (Adele Achem)
10                      Sandstrom, Miss. Marguerite Rut
11                             Bonnell, Miss. Elizabeth
12                       Saundercock, Mr. William Henry
13                          Andersson, Mr. Anders Johan
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

In [73]:
# get alternate records

In [74]:
titanic_df['Name'][0:15:2]

0                               Braund, Mr. Owen Harris
2                                Heikkinen, Miss. Laina
4                              Allen, Mr. William Henry
6                               McCarthy, Mr. Timothy J
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                      Sandstrom, Miss. Marguerite Rut
12                       Saundercock, Mr. William Henry
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

### check NULL values in Age column

In [75]:
# extract row index with null values in AGE column.

In [76]:
titanic_df[titanic_df['Age'].isnull() == True]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

In [77]:
# how many count of null values.

In [78]:
len(titanic_df[titanic_df['Age'].isnull() == True])

177

In [79]:
# show the entire record of 5th index that having null values. Use iloc.

In [80]:
titanic_df[titanic_df['Age'].isnull() == True].iloc[0: 1]

PassengerId  Survived  Pclass              Name   Sex  Age  SibSp  Parch  \
5            6         0       3  Moran, Mr. James  male  NaN      0      0   

   Ticket    Fare Cabin Embarked  
5  330877  8.4583   NaN        Q

In [81]:
# get records with higher fare

In [82]:
titanic_df[titanic_df['Fare'] == max(titanic_df['Fare'])]['Name']

258                      Ward, Miss. Anna
679    Cardeza, Mr. Thomas Drake Martinez
737                Lesurer, Mr. Gustave J
Name: Name, dtype: object

In [83]:
titanic_df['Fare'] == max(titanic_df['Fare'])

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Fare, Length: 891, dtype: bool

## Questions to Solve

In [84]:
# 1. how many male and female are passengers are on board

In [85]:
len(titanic_df[titanic_df['Sex']=='male'])

577

In [86]:
len(titanic_df[titanic_df['Sex']=='female'])

314

In [116]:
# Method 2 - Using value_counts

In [110]:
titanic_df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [115]:
# Method 3 - Using group by operation

In [114]:
titanic_df.groupby('Sex')['Sex'].count()

Sex
female    314
male      577
Name: Sex, dtype: int64

In [89]:
# 2. how many survived

In [117]:
len(titanic_df[titanic_df['Survived']==1])

342

In [121]:
# Method 2 - Using value_counts

In [122]:
titanic_df['Survived'].value_counts()[1]

342

In [123]:
# Method 3 - Using group by 

In [130]:
titanic_df.groupby('Survived')['Survived'].count()[1]

342

In [91]:
# 3. how many casualities

In [128]:
len(titanic_df[titanic_df['Survived']==0])

549

In [132]:
# Method 2 - Value Counts

In [133]:
titanic_df['Survived'].value_counts()[0]

549

In [134]:
# Method 3- group by 

In [131]:
titanic_df.groupby('Survived')['Survived'].count()[0]

549

In [93]:
# 4. find name of person who is the eldest one

In [135]:
titanic_df[titanic_df['Age'] == max(titanic_df['Age'])]['Name']

630    Barkworth, Mr. Algernon Henry Wilson
Name: Name, dtype: object

In [95]:
# 5. how many passengers we have in first, second and third class ?

In [96]:
len(titanic_df[titanic_df['Pclass']==1])

216

In [97]:
len(titanic_df[titanic_df['Pclass']==2])

184

In [98]:
len(titanic_df[titanic_df['Pclass']==3])

491

In [136]:
# method 2 - value counts

In [139]:
titanic_df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [140]:
# method 3- group by

In [141]:
titanic_df.groupby('Pclass')['Pclass'].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

In [99]:
# 6. no of person whose name starts with 's'

In [142]:
len(titanic_df[titanic_df['Name'].str.startswith('S') == True])

86

In [101]:
# 7. create new column with summation of 'Sibsp' and 'parch'

In [102]:
SibSp_Parch = titanic_df['SibSp'] + titanic_df['Parch']
titanic_df['Sum_SibSp_Parch'] = SibSp_Parch
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Sum_SibSp_Parch  
0      0         A/5 21171   7.2500   NaN        S                1  
1      0          PC 17599  71.2833   C85        C                1  
2      0  STON/O2. 3101282   7.9250   NaN        S                0

In [103]:
# 8. how many person with age less than 25.

In [104]:
len(titanic_df[titanic_df['Age']<25])

278

In [105]:
# 9. how many person died whose age below 25.

In [147]:
len(titanic_df[(titanic_df['Survived']==0) & (titanic_df['Age']<25)])

160

In [107]:
# 10. from a cabin column, seperate text and numeric value.

In [108]:
titanic_df[titanic_df['Cabin'].isnull()==True].head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3         Braund, Mr. Owen Harris    male   
2            3         1       3          Heikkinen, Miss. Laina  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0  22.0      1      0         A/5 21171   7.2500   NaN        S   
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
4  35.0      0      0            373450   8.0500   NaN        S   
5   NaN      0      0            330877   8.4583   NaN        Q   
7   2.0      3      1            349909  21.0750   NaN        S   

   Sum_SibSp_Parch  
0                1  
2                0  
4                0  
5                0  
7                4

In [109]:
titanic_df[titanic_df['Cabin'].isnull()==False].head()

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
6             7         0       1   
10           11         1       3   
11           12         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   

    Parch    Ticket     Fare Cabin Embarked  Sum_SibSp_Parch  
1       0  PC 17599  71.2833   C85        C                1  
3       0    113803  53.1000  C123        S                1  
6       0     17463  51.8625   E46        S                0  
10      1   PP 9549  16.7000    G6        S                2  
11      0    113783  26.5500  C103        S                0